Extract EditorMainScreen from EditorNode

kitbdev opened the pull request:
- related https://github.com/godotengine/godot/pull/87760

Extracts all functionality related to the main screen from EditorNode.
Behavior should not change.

### Details

In `_get_main_scene_state()`, removed the `main_tab` key. I don't think it has been used since 081a236.
In `_set_main_scene_state()`, removed the `editor_index` key check. Only used between https://github.com/godotengine/godot/commit/334a81844ee9f6d780dbd3c34f15863173b43e03 and #11075 (less than a month during 3.0 dev).
The state in only actually saved and used in `_save_central_editor_layout_to_config()` and `_load_central_editor_layout_from_config()`. `_get/set_main_scene_state()` is just used for switching between 2D/3D when opening a scene or changing selection to match the node.

Removed `distraction_free->move_to_front();` from `EditorNode::add_editor_plugin` since it isn't needed. The distraction free button is already after everything else in the SceneTabs, and if it was needed it should be somewhere else.

`set_visible_editor()` and `editor_select()` did the same thing. They are now `EditorMainScreen::select()`.
Renamed `editor_plugin_screen` -> `selected_plugin`, and removed some other `editor_main` prefixes.

The `scene_root_parent` is now the `editor_main_screen`. It is not actually the parent of the scene root, and this way I don't need to add a new control to hold the vbox.
I could make the `editor_main_screen` actually be the vbox instead of being parent to it, and the `scene_root_parent` can be called `main_screen_parent`, if that is better.

Also removed unused forward declarations in EditorNode.

kitbdev committed:
Extract editor main screen

KoBeWi commented on modules/csg/editor/csg_gizmos.cpp:
```diff
@@ -541,7 +542,7 @@ EditorPluginCSG::EditorPluginCSG() {
 
 	csg_shape_editor = memnew(CSGShapeEditor);
-	EditorNode::get_singleton()->get_main_screen_control()->add_child(csg_shape_editor);
+	EditorNode::get_singleton()->get_editor_main_screen()->get_control()->add_child(csg_shape_editor);
```
Editor's that don't have their screen should use `get_gui_base()` instead of main screen control.

KoBeWi commented on editor/export/editor_export_platform.cpp:
```diff
@@ -270,7 +271,7 @@ Error EditorExportPlatform::_save_zip_file(void *p_userdata, const String &p_pat
 	Ref<Theme> theme = EditorNode::get_singleton()->get_editor_theme();
 	ERR_FAIL_COND_V(theme.is_null(), Ref<ImageTexture>());
-	if (EditorNode::get_singleton()->get_main_screen_control()->is_layout_rtl()) {
+	if (EditorNode::get_singleton()->get_editor_main_screen()->get_control()->is_layout_rtl()) {
```
This should also be `get_gui_base()`.

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -5699,6 +5477,17 @@ void EditorNode::_toggle_distraction_free_mode() {
 }
 
+void EditorNode::update_distraction_free_mode() {
+	if (EDITOR_GET("interface/editor/separate_distraction_mode")) {
```
Can be early return.

KoBeWi commented on editor/editor_main_screen.cpp:
```diff
@@ -0,0 +1,300 @@
+void EditorMainScreen::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_READY: {
+			_select_default();
```
This is the only call here. It could be inlined, no reason to make a method.

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -7120,25 +6904,19 @@ EditorNode::EditorNode() {
-	scene_root_parent->set_v_size_flags(Control::SIZE_EXPAND_FILL);
+	editor_main_screen = memnew(EditorMainScreen);
+	editor_main_screen->set_custom_minimum_size(Size2(0, 80) * EDSCALE);
+	editor_main_screen->add_theme_style_override(SceneStringName(panel), theme->get_stylebox(SNAME("Content"), EditorStringName(EditorStyles)));
```
This line is called in `_update_theme()`, which is called at launch, so it can be removed from here.

KoBeWi commented:
Crash when clicking plain Node:
```
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.4.dev.custom_build (5675c76461e197d3929a1142cfb84ab1a76ac9dd)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[0] CowData<Button *>::get (C:\godot_source\core\templates\cowdata.h:205)
[1] Vector<Button *>::operator[] (C:\godot_source\core\templates\vector.h:97)
[2] EditorMainScreen::is_button_enabled (C:\godot_source\editor\editor_main_screen.cpp:118)
[3] EditorNode::_edit_current (C:\godot_source\editor\editor_node.cpp:2588)
[4] EditorNode::push_item (C:\godot_source\editor\editor_node.cpp:2345)
[5] EditorNode::push_node_item (C:\godot_source\editor\editor_node.cpp:2330)
[6] SceneTreeDock::_push_item (C:\godot_source\editor\scene_tree_dock.cpp:1769)
[7] SceneTreeDock::_post_do_create (C:\godot_source\editor\scene_tree_dock.cpp:2835)
[8] call_with_variant_args_helper<SceneTreeDock,Node *,0> (C:\godot_source\core\variant\binder_common.h:309)
[9] call_with_variant_args_dv<SceneTreeDock,Node *> (C:\godot_source\core\variant\binder_common.h:452)
[10] MethodBindT<SceneTreeDock,Node *>::call (C:\godot_source\core\object\method_bind.h:347)
[11] Object::callp (C:\godot_source\core\object\object.cpp:811)
[12] Callable::callp (C:\godot_source\core\variant\callable.cpp:69)
[13] CallableCustomBind::call (C:\godot_source\core\variant\callable_bind.cpp:153)
[14] Callable::callp (C:\godot_source\core\variant\callable.cpp:58)
[15] UndoRedo::_process_operation_list (C:\godot_source\core\object\undo_redo.cpp:366)
[16] UndoRedo::_redo (C:\godot_source\core\object\undo_redo.cpp:82)
[17] UndoRedo::commit_action (C:\godot_source\core\object\undo_redo.cpp:331)
[18] EditorUndoRedoManager::commit_action (C:\godot_source\editor\editor_undo_redo_manager.cpp:255)
[19] SceneTreeDock::_do_create (C:\godot_source\editor\scene_tree_dock.cpp:2826)
[20] SceneTreeDock::_create (C:\godot_source\editor\scene_tree_dock.cpp:2870)
[21] call_with_variant_args_helper<SceneTreeDock> (C:\godot_source\core\variant\binder_common.h:309)
[22] call_with_variant_args<SceneTreeDock> (C:\godot_source\core\variant\binder_common.h:419)
[23] CallableCustomMethodPointer<SceneTreeDock>::call (C:\godot_source\core\object\callable_method_pointer.h:104)
[24] Callable::callp (C:\godot_source\core\variant\callable.cpp:58)
[25] Object::emit_signalp (C:\godot_source\core\object\object.cpp:1199)
[26] Node::emit_signalp (C:\godot_source\scene\main\node.cpp:3927)
[27] Object::emit_signal<> (C:\godot_source\core\object\object.h:923)
[28] CreateDialog::_confirmed (C:\godot_source\editor\create_dialog.cpp:421)
[29] call_with_variant_args_helper<CreateDialog> (C:\godot_source\core\variant\binder_common.h:309)
[30] call_with_variant_args<CreateDialog> (C:\godot_source\core\variant\binder_common.h:419)
[31] CallableCustomMethodPointer<CreateDialog>::call (C:\godot_source\core\object\callable_method_pointer.h:104)
[32] Callable::callp (C:\godot_source\core\variant\callable.cpp:58)
[33] Object::emit_signalp (C:\godot_source\core\object\object.cpp:1199)
[34] Node::emit_signalp (C:\godot_source\scene\main\node.cpp:3927)
[35] Object::emit_signal<> (C:\godot_source\core\object\object.h:923)
[36] AcceptDialog::_ok_pressed (C:\godot_source\scene\gui\dialogs.cpp:138)
[37] AcceptDialog::_text_submitted (C:\godot_source\scene\gui\dialogs.cpp:124)
[38] call_with_variant_args_helper<AcceptDialog,String const &,0> (C:\godot_source\core\variant\binder_common.h:304)
[39] call_with_variant_args<AcceptDialog,String const &> (C:\godot_source\core\variant\binder_common.h:419)
[40] CallableCustomMethodPointer<AcceptDialog,String const &>::call (C:\godot_source\core\object\callable_method_pointer.h:104)
[41] Callable::callp (C:\godot_source\core\variant\callable.cpp:58)
[42] Object::emit_signalp (C:\godot_source\core\object\object.cpp:1199)
[43] Node::emit_signalp (C:\godot_source\scene\main\node.cpp:3927)
[44] Object::emit_signal<String> (C:\godot_source\core\object\object.h:923)
[45] LineEdit::gui_input (C:\godot_source\scene\gui\line_edit.cpp:523)
[46] Control::_call_gui_input (C:\godot_source\scene\gui\control.cpp:1829)
[47] Viewport::_gui_input_event (C:\godot_source\scene\main\viewport.cpp:2216)
[48] Viewport::push_input (C:\godot_source\scene\main\viewport.cpp:3267)
[49] Window::_window_input (C:\godot_source\scene\main\window.cpp:1682)
[50] call_with_variant_args_helper<Window,Ref<InputEvent> const &,0> (C:\godot_source\core\variant\binder_common.h:304)
[51] call_with_variant_args<Window,Ref<InputEvent> const &> (C:\godot_source\core\variant\binder_common.h:419)
[52] CallableCustomMethodPointer<Window,Ref<InputEvent> const &>::call (C:\godot_source\core\object\callable_method_pointer.h:104)
[53] Callable::callp (C:\godot_source\core\variant\callable.cpp:58)
[54] Callable::call<Ref<InputEvent> > (C:\godot_source\core\variant\variant.h:875)
[55] DisplayServerWindows::_dispatch_input_event (C:\godot_source\platform\windows\display_server_windows.cpp:3733)
[56] DisplayServerWindows::_dispatch_input_events (C:\godot_source\platform\windows\display_server_windows.cpp:3704)
[57] Input::_parse_input_event_impl (C:\godot_source\core\input\input.cpp:803)
[58] Input::flush_buffered_events (C:\godot_source\core\input\input.cpp:1084)
[59] DisplayServerWindows::process_events (C:\godot_source\platform\windows\display_server_windows.cpp:3186)
[60] OS_Windows::run (C:\godot_source\platform\windows\os_windows.cpp:1772)
[61] widechar_main (C:\godot_source\platform\windows\godot_windows.cpp:181)
[62] _main (C:\godot_source\platform\windows\godot_windows.cpp:206)
[63] main (C:\godot_source\platform\windows\godot_windows.cpp:220)
[64] WinMain (C:\godot_source\platform\windows\godot_windows.cpp:234)
[65] __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[66] <couldn't map PC to fn name>
-- END OF BACKTRACE --
```

kitbdev commented:
Fixed crash, it was because I used `get_plugin_index` which can return -1 but I didn't check for it. Also added `ERR_FAIL_INDEX_V` so it won't crash if it happens again.

Only plugins with a main screen are now added as children to the main screen control, otherwise they use gui_base.

KoBeWi commented on editor/editor_main_screen.cpp:
```diff
@@ -0,0 +1,298 @@
+}
+
+void EditorMainScreen::load_layout_from_config(Ref<ConfigFile> p_config_file, const String &p_section) {
+	if (p_config_file->has_section_key(p_section, "selected_main_editor_idx")) {
+		int selected_main_editor_idx = p_config_file->get_value(p_section, "selected_main_editor_idx");
```
Could be simplified to
```suggestion
	int selected_main_editor_idx = p_config_file->get_value(p_section, "selected_main_editor_idx", -1);
```

KoBeWi commented on editor/editor_main_screen.cpp:
```diff
@@ -0,0 +1,298 @@
+	selecting = true;
+
+	for (int i = 0; i < buttons.size(); i++) {
+		buttons[i]->set_pressed(i == p_index);
```
```suggestion
		buttons[i]->set_pressed_no_signal(i == p_index);
```
Allows to remove `selecting`.

KoBeWi reviewed (approved):
Looks fine.
I noticed some potential improvements that could be done in a follow-up, e.g. merging `buttons` and `editor_table` by using a struct.

akien-mga commented:
Thanks!
